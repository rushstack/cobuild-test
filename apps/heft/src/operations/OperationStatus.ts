// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * Enumeration defining potential states of an operation
 * @beta
 */
export enum OperationStatus {
  /**
   * The Operation is on the queue, ready to execute (but may be waiting for dependencies)
   */
  Ready = 'READY',
  /**
   * The Operation is currently executing
   */
  Executing = 'EXECUTING',
  /**
   * The Operation completed successfully and did not write to standard output
   */
  Success = 'SUCCESS',
  /**
   * The Operation failed
   */
  Failure = 'FAILURE',
  /**
   * The operation was cancelled
   */
  Cancelled = 'CANCELLED',
  /**
   * The Operation could not be executed because one or more of its dependencies failed
   */
  Blocked = 'BLOCKED',
  /**
   * The operation performed no meaningful work.
   */
  NoOp = 'NO OP'
}
