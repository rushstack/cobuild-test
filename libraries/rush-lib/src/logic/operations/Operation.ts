// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { RushConfigurationProject } from '../../api/RushConfigurationProject';
import { IPhase } from '../../api/CommandLineConfiguration';
import { IOperationRunner } from './IOperationRunner';

/**
 * Options for constructing a new Operation.
 * @alpha
 */
export interface IOperationOptions {
  /**
   * The Rush phase associated with this Operation, if any
   */
  phase?: IPhase | undefined;
  /**
   * The Rush project associated with this Operation, if any
   */
  project?: RushConfigurationProject | undefined;
  /**
   * When the scheduler is ready to process this `Operation`, the `runner` implements the actual work of
   * running the operation.
   */
  runner?: IOperationRunner | undefined;
}

/**
 * The `Operation` class is a node in the dependency graph of work that needs to be scheduled by the
 * `OperationExecutionManager`. Each `Operation` has a `runner` member of type `IOperationRunner`, whose
 * implementation manages the actual process of running a single operation.
 *
 * The graph of `Operation` instances will be cloned into a separate execution graph after processing.
 *
 * @alpha
 */
export class Operation {
  /**
   * The Rush phase associated with this Operation, if any
   */
  public readonly associatedPhase: IPhase | undefined;

  /**
   * The Rush project associated with this Operation, if any
   */
  public readonly associatedProject: RushConfigurationProject | undefined;

  /**
   * A set of all operations which depend on this operation.
   */
  public readonly consumers: ReadonlySet<Operation> = new Set<Operation>();

  /**
   * A set of all dependencies which must be executed before this operation is complete.
   */
  public readonly dependencies: ReadonlySet<Operation> = new Set<Operation>();

  /**
   * When the scheduler is ready to process this `Operation`, the `runner` implements the actual work of
   * running the operation.
   */
  public runner: IOperationRunner | undefined = undefined;

  /**
   * The weight for this operation. This scalar is the contribution of this operation to the
   * `criticalPathLength` calculation above. Modify to indicate the following:
   * - `weight` === 1: indicates that this operation has an average duration
   * - `weight` &gt; 1: indicates that this operation takes longer than average and so the scheduler
   *     should try to favor starting it over other, shorter operations. An example might be an operation that
   *     bundles an entire application and runs whole-program optimization.
   * - `weight` &lt; 1: indicates that this operation takes less time than average and so the scheduler
   *     should favor other, longer operations over it. An example might be an operation to unpack a cached
   *     output, or an operation using NullOperationRunner, which might use a value of 0.
   */
  public weight: number = 1;

  public constructor(options?: IOperationOptions) {
    this.associatedPhase = options?.phase;
    this.associatedProject = options?.project;
    this.runner = options?.runner;
  }

  /**
   * The name of this operation, for logging.
   */
  public get name(): string | undefined {
    return this.runner?.name;
  }

  /**
   * Adds the specified operation as a dependency and updates the consumer list.
   */
  public addDependency(dependency: Operation): void {
    // Cast internally to avoid adding the overhead of getters
    (this.dependencies as Set<Operation>).add(dependency);
    (dependency.consumers as Set<Operation>).add(this);
  }

  /**
   * Deletes the specified operation as a dependency and updates the consumer list.
   */
  public deleteDependency(dependency: Operation): void {
    // Cast internally to avoid adding the overhead of getters
    (this.dependencies as Set<Operation>).delete(dependency);
    (dependency.consumers as Set<Operation>).delete(this);
  }
}
