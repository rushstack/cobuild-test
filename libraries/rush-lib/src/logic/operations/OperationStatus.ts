// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * Enumeration defining potential states of an operation
 * @beta
 */
export enum OperationStatus {
  /**
   * The Operation is on the queue, ready to execute (but may be waiting for dependencies)
   */
  Ready = 'READY',
  /**
   * The Operation is Queued
   */
  Queued = 'QUEUED',
  /**
   * The Operation is currently executing
   */
  Executing = 'EXECUTING',
  /**
   * The Operation is currently executing by a remote process
   */
  RemoteExecuting = 'REMOTE EXECUTING',
  /**
   * The Operation completed successfully and did not write to standard output
   */
  Success = 'SUCCESS',
  /**
   * The Operation completed successfully, but wrote to standard output
   */
  SuccessWithWarning = 'SUCCESS WITH WARNINGS',
  /**
   * The Operation was skipped via the legacy incremental build logic
   */
  Skipped = 'SKIPPED',
  /**
   * The Operation had its outputs restored from the build cache
   */
  FromCache = 'FROM CACHE',
  /**
   * The Operation failed
   */
  Failure = 'FAILURE',
  /**
   * The Operation could not be executed because one or more of its dependencies failed
   */
  Blocked = 'BLOCKED',
  /**
   * The Operation was a no-op (for example, it had an empty script)
   */
  NoOp = 'NO OP'
}
