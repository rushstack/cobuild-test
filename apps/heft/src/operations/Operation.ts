// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { InternalError, ITerminal } from '@rushstack/node-core-library';

import { Stopwatch } from '../utilities/Stopwatch';
import type {
  IOperationRunner,
  IOperationRunnerContext,
  IOperationState,
  IOperationStates
} from './IOperationRunner';
import { OperationError } from './OperationError';
import { OperationStatus } from './OperationStatus';

/**
 * Options for constructing a new Operation.
 * @alpha
 */
export interface IOperationOptions {
  /**
   * The group that this operation belongs to. Will be used for logging and duration tracking.
   */
  groupName?: string | undefined;

  /**
   * When the scheduler is ready to process this `Operation`, the `runner` implements the actual work of
   * running the operation.
   */
  runner?: IOperationRunner | undefined;

  /**
   * The weight used by the scheduler to determine order of execution.
   */
  weight?: number | undefined;
}

/**
 * Information provided to `executeAsync` by the `OperationExecutionManager`.
 */
export interface IExecuteOperationContext extends Omit<IOperationRunnerContext, 'isFirstRun' | 'requestRun'> {
  /**
   * Function to invoke before execution of an operation, for logging.
   */
  beforeExecute(operation: Operation, state: IOperationState): void;
  /**
   * Function to invoke after execution of an operation, for logging.
   */
  afterExecute(operation: Operation, state: IOperationState): void;
  /**
   * Function used to schedule the concurrency-limited execution of an operation.
   */
  queueWork<T>(workFn: () => Promise<T>, priority: number): Promise<T>;

  /**
   * A callback to the overarching orchestrator to request that the operation be invoked again.
   * Used in watch mode to signal that inputs have changed.
   */
  requestRun?: (requestor?: string) => void;

  /**
   * Terminal to write output to.
   */
  terminal: ITerminal;
}

/**
 * The `Operation` class is a node in the dependency graph of work that needs to be scheduled by the
 * `OperationExecutionManager`. Each `Operation` has a `runner` member of type `IOperationRunner`, whose
 * implementation manages the actual process of running a single operation.
 *
 * The graph of `Operation` instances will be cloned into a separate execution graph after processing.
 *
 * @alpha
 */
export class Operation implements IOperationStates {
  /**
   * A set of all dependencies which must be executed before this operation is complete.
   */
  public readonly dependencies: Set<Operation> = new Set<Operation>();
  /**
   * A set of all operations that wait for this operation.
   */
  public readonly consumers: Set<Operation> = new Set<Operation>();
  /**
   * If specified, the name of a grouping to which this Operation belongs, for logging start and end times.
   */
  public readonly groupName: string | undefined;

  /**
   * When the scheduler is ready to process this `Operation`, the `runner` implements the actual work of
   * running the operation.
   */
  public runner: IOperationRunner | undefined = undefined;

  /**
   * This number represents how far away this Operation is from the furthest "root" operation (i.e.
   * an operation with no consumers). This helps us to calculate the critical path (i.e. the
   * longest chain of projects which must be executed in order, thereby limiting execution speed
   * of the entire operation tree.
   *
   * This number is calculated via a memoized depth-first search, and when choosing the next
   * operation to execute, the operation with the highest criticalPathLength is chosen.
   *
   * Example:
   *        (0) A
   *             \
   *          (1) B     C (0)         (applications)
   *               \   /|\
   *                \ / | \
   *             (2) D  |  X (1)      (utilities)
   *                    | / \
   *                    |/   \
   *                (2) Y     Z (2)   (other utilities)
   *
   * All roots (A & C) have a criticalPathLength of 0.
   * B has a score of 1, since A depends on it.
   * D has a score of 2, since we look at the longest chain (e.g D->B->A is longer than D->C)
   * X has a score of 1, since the only package which depends on it is A
   * Z has a score of 2, since only X depends on it, and X has a score of 1
   * Y has a score of 2, since the chain Y->X->C is longer than Y->C
   *
   * The algorithm is implemented in AsyncOperationQueue.ts as calculateCriticalPathLength()
   */
  public criticalPathLength: number | undefined = undefined;

  /**
   * The weight for this operation. This scalar is the contribution of this operation to the
   * `criticalPathLength` calculation above. Modify to indicate the following:
   * - `weight` === 1: indicates that this operation has an average duration
   * - `weight` &gt; 1: indicates that this operation takes longer than average and so the scheduler
   *     should try to favor starting it over other, shorter operations. An example might be an operation that
   *     bundles an entire application and runs whole-program optimization.
   * - `weight` &lt; 1: indicates that this operation takes less time than average and so the scheduler
   *     should favor other, longer operations over it. An example might be an operation to unpack a cached
   *     output, or an operation using NullOperationRunner, which might use a value of 0.
   */
  public weight: number;

  /**
   * The state of this operation the previous time a manager was invoked.
   */
  public lastState: IOperationState | undefined = undefined;

  /**
   * The current state of this operation
   */
  public state: IOperationState | undefined = undefined;

  /**
   * A cached execution promise for the current OperationExecutionManager invocation of this operation.
   */
  private _promise: Promise<OperationStatus> | undefined = undefined;

  /**
   * If true, then a run of this operation is currently wanted.
   * This is used to track state from the `requestRun` callback passed to the runner.
   */
  private _runPending: boolean = true;

  public constructor(options?: IOperationOptions) {
    this.groupName = options?.groupName;
    this.runner = options?.runner;
    this.weight = options?.weight || 1;
  }

  /**
   * The name of this operation, for logging.
   */
  public get name(): string | undefined {
    return this.runner?.name;
  }

  public addDependency(dependency: Operation): void {
    this.dependencies.add(dependency);
    dependency.consumers.add(this);
  }

  public deleteDependency(dependency: Operation): void {
    this.dependencies.delete(dependency);
    dependency.consumers.delete(this);
  }

  public reset(): void {
    // Reset operation state
    this.lastState = this.state;

    this.state = {
      status: OperationStatus.Ready,
      error: undefined,
      stopwatch: new Stopwatch()
    };

    this._promise = undefined;
    this._runPending = true;
  }

  /**
   * @internal
   */
  public async _executeAsync(context: IExecuteOperationContext): Promise<OperationStatus> {
    const { state } = this;
    if (!state) {
      throw new Error(`Operation state has not been initialized.`);
    }

    if (!this._promise) {
      this._promise = this._executeInnerAsync(context, state);
    }

    return this._promise;
  }

  private async _executeInnerAsync(
    context: IExecuteOperationContext,
    rawState: IOperationState
  ): Promise<OperationStatus> {
    const state: IOperationState = rawState;
    const { runner } = this;

    const dependencyResults: PromiseSettledResult<OperationStatus>[] = await Promise.allSettled(
      Array.from(this.dependencies, (dependency: Operation) => dependency._executeAsync(context))
    );

    const { cancellationToken, requestRun, queueWork } = context;

    if (cancellationToken.isCancelled) {
      state.status = OperationStatus.Cancelled;
      return state.status;
    }

    for (const result of dependencyResults) {
      if (
        result.status === 'rejected' ||
        result.value === OperationStatus.Blocked ||
        result.value === OperationStatus.Failure
      ) {
        state.status = OperationStatus.Blocked;
        return state.status;
      }
    }

    const innerContext: IOperationRunnerContext = {
      cancellationToken,
      isFirstRun: !this.lastState,
      requestRun: requestRun
        ? () => {
            switch (this.state?.status) {
              case OperationStatus.Ready:
              case OperationStatus.Executing:
                // If current status has not yet resolved to a fixed value,
                // re-executing this operation does not require a full rerun
                // of the operation graph. Simply mark that a run is requested.

                // This variable is on the Operation instead of the
                // containing closure to deal with scenarios in which
                // the runner hangs on to an old copy of the callback.
                this._runPending = true;
                return;

              case OperationStatus.Blocked:
              case OperationStatus.Cancelled:
              case OperationStatus.Failure:
              case OperationStatus.NoOp:
              case OperationStatus.Success:
                // The requestRun callback is assumed to remain constant
                // throughout the lifetime of the process, so it is safe
                // to capture here.
                return requestRun(this.name);
              default:
                throw new InternalError(`Unexpected status: ${this.state?.status}`);
            }
          }
        : undefined
    };

    await queueWork(async () => {
      // Redundant variable to satisfy require-atomic-updates
      const innerState: IOperationState = state;

      if (cancellationToken.isCancelled) {
        innerState.status = OperationStatus.Cancelled;
        return innerState.status;
      }

      context.beforeExecute(this, innerState);

      innerState.status = OperationStatus.Executing;
      innerState.stopwatch.start();

      while (this._runPending) {
        this._runPending = false;
        try {
          innerState.status = runner ? await runner.executeAsync(innerContext) : OperationStatus.NoOp;
        } catch (error) {
          innerState.status = OperationStatus.Failure;
          innerState.error = error as OperationError;
        }

        // Since runner.executeAsync is async, a change could have occurred that requires re-execution
        // This operation is still active, so can re-execute immediately, rather than forcing a whole
        // new execution pass.

        // As currently written, this does mean that if a job is scheduled with higher priority while
        // this operation is still executing, it will still wait for this retry. This may not be desired
        // and if it becomes a problem, the retry loop will need to be moved outside of the `queueWork` call.
        // This introduces complexity regarding tracking of timing and start/end logging, however.

        if (this._runPending) {
          if (cancellationToken.isCancelled) {
            innerState.status = OperationStatus.Cancelled;
            break;
          } else {
            context.terminal.writeLine(`Immediate rerun requested. Executing.`);
          }
        }
      }

      state.stopwatch.stop();
      context.afterExecute(this, state);
    }, this.criticalPathLength ?? 0);

    return state.status;
  }
}
