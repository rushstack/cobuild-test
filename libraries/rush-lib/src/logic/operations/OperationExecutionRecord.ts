// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { StdioSummarizer } from '@rushstack/terminal';
import { InternalError } from '@rushstack/node-core-library';
import { CollatedWriter, StreamCollator } from '@rushstack/stream-collator';

import { OperationStatus } from './OperationStatus';
import { IOperationRunner, IOperationRunnerContext } from './IOperationRunner';
import { Operation } from './Operation';
import { Stopwatch } from '../../utilities/Stopwatch';
import { OperationMetadataManager } from './OperationMetadataManager';
import type { IPhase } from '../../api/CommandLineConfiguration';
import type { RushConfigurationProject } from '../../api/RushConfigurationProject';

export interface IOperationExecutionRecordContext {
  streamCollator: StreamCollator;
  onOperationStatusChanged?: (record: OperationExecutionRecord) => void;

  debugMode: boolean;
  quietMode: boolean;
  changedProjectsOnly: boolean;
}

/**
 * Internal class representing everything about executing an operation
 *
 * @internal
 */
export class OperationExecutionRecord implements IOperationRunnerContext {
  /**
   * The associated operation.
   */
  public readonly operation: Operation;

  /**
   * The current execution status of an operation. Operations start in the 'ready' state,
   * but can be 'blocked' if an upstream operation failed. It is 'executing' when
   * the operation is executing. Once execution is complete, it is either 'success' or
   * 'failure'.
   */
  public status: OperationStatus = OperationStatus.Ready;

  /**
   * The error which occurred while executing this operation, this is stored in case we need
   * it later (for example to re-print errors at end of execution).
   */
  public error: Error | undefined = undefined;

  /**
   * This number represents how far away this Operation is from the furthest "root" operation (i.e.
   * an operation with no consumers). This helps us to calculate the critical path (i.e. the
   * longest chain of projects which must be executed in order, thereby limiting execution speed
   * of the entire operation tree.
   *
   * This number is calculated via a memoized depth-first search, and when choosing the next
   * operation to execute, the operation with the highest criticalPathLength is chosen.
   *
   * Example:
   * ```
   *        (0) A
   *             \
   *          (1) B     C (0)         (applications)
   *               \   /|\
   *                \ / | \
   *             (2) D  |  X (1)      (utilities)
   *                    | / \
   *                    |/   \
   *                (2) Y     Z (2)   (other utilities)
   *
   * All roots (A & C) have a criticalPathLength of 0.
   * B has a score of 1, since A depends on it.
   * D has a score of 2, since we look at the longest chain (e.g D->B->A is longer than D->C)
   * X has a score of 1, since the only package which depends on it is A
   * Z has a score of 2, since only X depends on it, and X has a score of 1
   * Y has a score of 2, since the chain Y->X->C is longer than Y->C
   * ```
   *
   * The algorithm is implemented in AsyncOperationQueue.ts as calculateCriticalPathLength()
   */
  public criticalPathLength: number | undefined = undefined;

  /**
   * The set of operations that must complete before this operation executes.
   */
  public readonly dependencies: Set<OperationExecutionRecord> = new Set();
  /**
   * The set of operations that depend on this operation.
   */
  public readonly consumers: Set<OperationExecutionRecord> = new Set();

  public readonly stopwatch: Stopwatch = new Stopwatch();
  public readonly stdioSummarizer: StdioSummarizer = new StdioSummarizer();

  public readonly runner: IOperationRunner;
  public readonly weight: number;
  public readonly associatedPhase: IPhase | undefined;
  public readonly associatedProject: RushConfigurationProject | undefined;
  public readonly _operationMetadataManager: OperationMetadataManager | undefined;

  private readonly _context: IOperationExecutionRecordContext;

  private _collatedWriter: CollatedWriter | undefined = undefined;

  public constructor(operation: Operation, context: IOperationExecutionRecordContext) {
    const { runner, associatedPhase, associatedProject } = operation;

    if (!runner) {
      throw new InternalError(
        `Operation for phase '${associatedPhase?.name}' and project '${associatedProject?.packageName}' has no runner.`
      );
    }

    this.operation = operation;
    this.runner = runner;
    this.weight = operation.weight;
    this.associatedPhase = associatedPhase;
    this.associatedProject = associatedProject;
    if (operation.associatedPhase && operation.associatedProject) {
      this._operationMetadataManager = new OperationMetadataManager({
        phase: operation.associatedPhase,
        rushProject: operation.associatedProject
      });
    }
    this._context = context;
  }

  public get name(): string {
    return this.runner.name;
  }

  public get debugMode(): boolean {
    return this._context.debugMode;
  }

  public get quietMode(): boolean {
    return this._context.quietMode;
  }

  public get changedProjectsOnly(): boolean {
    return this._context.changedProjectsOnly;
  }

  public get collatedWriter(): CollatedWriter {
    // Lazy instantiate because the registerTask() call affects display ordering
    if (!this._collatedWriter) {
      this._collatedWriter = this._context.streamCollator.registerTask(this.name);
    }
    return this._collatedWriter;
  }

  public get nonCachedDurationMs(): number | undefined {
    // Lazy calculated because the state file is created/restored later on
    return this._operationMetadataManager?.stateFile.state?.nonCachedDurationMs;
  }

  public get cobuildRunnerId(): string | undefined {
    // Lazy calculated because the state file is created/restored later on
    return this._operationMetadataManager?.stateFile.state?.cobuildRunnerId;
  }

  public async executeAsync({
    onStart,
    onResult
  }: {
    onStart: (record: OperationExecutionRecord) => Promise<OperationStatus | undefined>;
    onResult: (record: OperationExecutionRecord) => Promise<void>;
  }): Promise<void> {
    if (this.status === OperationStatus.RemoteExecuting) {
      this.stopwatch.reset();
    }
    this.status = OperationStatus.Executing;
    this.stopwatch.start();
    this._context.onOperationStatusChanged?.(this);

    try {
      const earlyReturnStatus: OperationStatus | undefined = await onStart(this);
      // When the operation status returns by the hook, bypass the runner execution.
      if (earlyReturnStatus) {
        this.status = earlyReturnStatus;
      } else {
        this.status = await this.runner.executeAsync(this);
      }
      // Delegate global state reporting
      await onResult(this);
    } catch (error) {
      this.status = OperationStatus.Failure;
      this.error = error;
      // Delegate global state reporting
      await onResult(this);
    } finally {
      if (this.status !== OperationStatus.RemoteExecuting) {
        this._collatedWriter?.close();
        this.stdioSummarizer.close();
        this.stopwatch.stop();
      }
      this._context.onOperationStatusChanged?.(this);
    }
  }
}
