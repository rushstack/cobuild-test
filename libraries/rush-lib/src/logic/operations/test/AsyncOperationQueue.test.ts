// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { Operation } from '../Operation';
import { IOperationExecutionRecordContext, OperationExecutionRecord } from '../OperationExecutionRecord';
import { MockOperationRunner } from './MockOperationRunner';
import {
  AsyncOperationQueue,
  IOperationIteratorResult,
  IOperationSortFunction,
  UNASSIGNED_OPERATION
} from '../AsyncOperationQueue';
import { OperationStatus } from '../OperationStatus';
import { Async } from '@rushstack/node-core-library';

function addDependency(consumer: OperationExecutionRecord, dependency: OperationExecutionRecord): void {
  consumer.dependencies.add(dependency);
  dependency.consumers.add(consumer);
}

function nullSort(a: OperationExecutionRecord, b: OperationExecutionRecord): number {
  return 0;
}

function createRecord(name: string): OperationExecutionRecord {
  return new OperationExecutionRecord(
    new Operation({
      runner: new MockOperationRunner(name)
    }),
    {} as unknown as IOperationExecutionRecordContext
  );
}

describe(AsyncOperationQueue.name, () => {
  it('iterates operations in topological order', async () => {
    const operations = [createRecord('a'), createRecord('b'), createRecord('c'), createRecord('d')];

    addDependency(operations[0], operations[2]);
    addDependency(operations[3], operations[1]);
    addDependency(operations[1], operations[0]);

    const expectedOrder = [operations[2], operations[0], operations[1], operations[3]];
    const actualOrder = [];
    // Nothing sets the RemoteExecuting status, this should be a error if it happens
    let hasUnassignedOperation: boolean = false;
    const queue: AsyncOperationQueue = new AsyncOperationQueue(operations, nullSort);
    for await (const operation of queue) {
      actualOrder.push(operation);
      if (operation === UNASSIGNED_OPERATION) {
        hasUnassignedOperation = true;
        continue;
      }
      for (const consumer of operation.consumers) {
        consumer.dependencies.delete(operation);
      }
      operation.status = OperationStatus.Success;
      queue.complete(operation);
    }

    expect(actualOrder).toEqual(expectedOrder);
    expect(hasUnassignedOperation).toEqual(false);
  });

  it('respects the sort predicate', async () => {
    const operations = [createRecord('a'), createRecord('b'), createRecord('c'), createRecord('d')];

    const expectedOrder = [operations[2], operations[0], operations[1], operations[3]];
    const actualOrder = [];
    const customSort: IOperationSortFunction = (
      a: OperationExecutionRecord,
      b: OperationExecutionRecord
    ): number => {
      return expectedOrder.indexOf(b) - expectedOrder.indexOf(a);
    };
    // Nothing sets the RemoteExecuting status, this should be a error if it happens
    let hasUnassignedOperation: boolean = false;

    const queue: AsyncOperationQueue = new AsyncOperationQueue(operations, customSort);
    for await (const operation of queue) {
      actualOrder.push(operation);
      if (operation === UNASSIGNED_OPERATION) {
        hasUnassignedOperation = true;
        continue;
      }
      for (const consumer of operation.consumers) {
        consumer.dependencies.delete(operation);
      }
      operation.status = OperationStatus.Success;
      queue.complete(operation);
    }

    expect(actualOrder).toEqual(expectedOrder);

    expect(hasUnassignedOperation).toEqual(false);
  });

  it('detects cycles', async () => {
    const operations = [createRecord('a'), createRecord('b'), createRecord('c'), createRecord('d')];

    addDependency(operations[0], operations[2]);
    addDependency(operations[2], operations[3]);
    addDependency(operations[3], operations[1]);
    addDependency(operations[1], operations[0]);

    expect(() => {
      new AsyncOperationQueue(operations, nullSort);
    }).toThrowErrorMatchingSnapshot();
  });

  it('handles concurrent iteration', async () => {
    const operations = [
      createRecord('a'),
      createRecord('b'),
      createRecord('c'),
      createRecord('d'),
      createRecord('e')
    ];

    // Set up to allow (0,1) -> (2) -> (3,4)
    addDependency(operations[2], operations[0]);
    addDependency(operations[2], operations[1]);
    addDependency(operations[3], operations[2]);
    addDependency(operations[4], operations[2]);

    const expectedConcurrency = new Map([
      [operations[0], 2],
      [operations[1], 2],
      [operations[2], 1],
      [operations[3], 2],
      [operations[4], 2]
    ]);

    const actualConcurrency: Map<OperationExecutionRecord, number> = new Map();
    const queue: AsyncOperationQueue = new AsyncOperationQueue(operations, nullSort);
    let concurrency: number = 0;
    // Nothing sets the RemoteExecuting status, this should be a error if it happens
    let hasUnassignedOperation: boolean = false;

    // Use 3 concurrent iterators to verify that it handles having more than the operation concurrency
    await Promise.all(
      Array.from({ length: 3 }, async () => {
        for await (const operation of queue) {
          if (operation === UNASSIGNED_OPERATION) {
            hasUnassignedOperation = true;
            continue;
          }
          ++concurrency;
          await Promise.resolve();

          actualConcurrency.set(operation, concurrency);

          await Promise.resolve();

          for (const consumer of operation.consumers) {
            consumer.dependencies.delete(operation);
          }

          --concurrency;
          operation.status = OperationStatus.Success;
          queue.complete(operation);
        }
      })
    );

    for (const [operation, operationConcurrency] of expectedConcurrency) {
      expect(actualConcurrency.get(operation)).toEqual(operationConcurrency);
    }

    expect(hasUnassignedOperation).toEqual(false);
  });

  it('handles remote executed operations', async () => {
    const operations = [
      createRecord('a'),
      createRecord('b'),
      createRecord('c'),
      createRecord('d'),
      createRecord('e')
    ];

    addDependency(operations[2], operations[1]);
    addDependency(operations[3], operations[1]);
    addDependency(operations[4], operations[1]);
    addDependency(operations[3], operations[2]);
    addDependency(operations[4], operations[3]);

    // b remote executing -> a -> b (remote executed) -> c -> d -> e
    const expectedOrder: string[] = ['b', 'a', 'b', 'c', 'd', 'e'];

    const queue: AsyncOperationQueue = new AsyncOperationQueue(operations, nullSort);

    const actualOrder: string[] = [];
    let remoteExecuted: boolean = false;
    for await (const operation of queue) {
      let record: OperationExecutionRecord | undefined;
      if (operation === UNASSIGNED_OPERATION) {
        await Async.sleep(100);
        record = queue.tryGetRemoteExecutingOperation();
      } else {
        record = operation;
      }
      if (!record) {
        continue;
      }

      actualOrder.push(record.name);

      if (record === operations[1]) {
        if (!remoteExecuted) {
          operations[1].status = OperationStatus.RemoteExecuting;
          // remote executed operation is finished later
          remoteExecuted = true;
          continue;
        }
      }
      for (const consumer of record.consumers) {
        consumer.dependencies.delete(record);
      }
      record.status = OperationStatus.Success;
      queue.complete(record);
    }

    expect(actualOrder).toEqual(expectedOrder);
  });

  it('handles an empty queue', async () => {
    const operations: OperationExecutionRecord[] = [];

    const queue: AsyncOperationQueue = new AsyncOperationQueue(operations, nullSort);
    const iterator: AsyncIterator<IOperationIteratorResult> = queue[Symbol.asyncIterator]();
    const result: IteratorResult<IOperationIteratorResult> = await iterator.next();
    expect(result.done).toEqual(true);
  });
});
