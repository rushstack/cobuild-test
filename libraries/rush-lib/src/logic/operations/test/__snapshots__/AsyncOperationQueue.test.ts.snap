// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncOperationQueue detects cycles 1`] = `
"A cyclic dependency was encountered:
  a
  -> c
  -> d
  -> b
  -> a
Consider using the decoupledLocalDependencies option for rush.json."
`;
